### PR TITLE
Data aggregation scaling study

### PR DESCRIPTION
**Pull request by mellis13**
This PR adds a data aggregation synthetic scaling study.  This scaling study has a producer and consumer application for testing aggregation API throughput.  The scaling study allows for the number of threads for dataset aggregation to be changed programmatically via the driver script.

### CODE REVIEW

File: cpp-data-aggregation/aggregation_producer.cpp [hunk 1]
```diff
@@ -0,0 +1,122 @@
+#include "client.h"
+#include <mpi.h>
+
+
+int get_iterations() {
+  char* iterations = std::getenv("SS_ITERATIONS");
+  int iters = iterations ? std::stoi(iterations) : 100;
```
Comment (ashao) [inline cpp-data-aggregation/aggregation_producer.cpp:7]: Same as above

File: cpp-data-aggregation/aggregation_consumer.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+    // Block to make sure all clients are connected
+    MPI_Barrier(MPI_COMM_WORLD);
+
+    // Retrieve rank-local loop start time
+    double loop_start = MPI_Wtime();
+
+    // Perform dataset aggregation retrieval
+    for (int i=0; i<iterations; i++) {
+
```
Comment (ashao) [inline cpp-data-aggregation/aggregation_consumer.cpp:42]: I'd suggest putting in an `MPI_Barrier` here just on the off chance that `client.delete_list` takes forever to return. At the very least we'll know that all ranks are perfectly synced at the start of the iterations

File: driver.py [hunk 3]
```diff
@@ -594,6 +709,143 @@ def create_throughput_session(exp,
                 tensor_bytes=_bytes)
     return model
 
+def create_aggregation_prod_session(exp,
```
Comment (ashao) [inline driver.py:712]: Slight preference to spell out producer instead of `prod`

File: driver.py [hunk 4]
```diff
@@ -594,6 +709,143 @@ def create_throughput_session(exp,
...
+                client_per_node=tasks,
+                client_nodes=nodes,
+                database_nodes=db_nodes,
+                database_cpus=db_cpus,
+                iterations=iterations,
+                tensor_bytes=_bytes,
+                t_per_dataset=t_per_dataset)
+    return model
+
+def create_aggregation_cons_session(exp,
```
Comment (ashao) [inline driver.py:776]: `consumer` for `cons`

File: driver.py [hunk 5]
```diff
@@ -594,6 +709,143 @@ def create_throughput_session(exp,
...
+    :param c_threads:  the number of client threads to use inside of SR client
+    :rtype c_threads: int
+    :type t_per_dataset: int
+    :rtype: Model
+    """
+    settings = exp.create_run_settings("./cpp-data-aggregation/build/aggregation_consumer",
+                                       [str(nodes*tasks)])
+    #settings.set_tasks(1)
+    settings.set_tasks_per_node(1)
+    settings.set_cpus_per_task(c_threads*2)
```
Comment (ashao) [inline driver.py:812]: The `*2` is for hyperthreads? If so, for systems that have HT disabled, this should be controlled by an optional parameter.

File: driver.py [hunk 6]
```diff
@@ -298,6 +298,121 @@ def throughput_scaling(self,
...
+                            clients_per_node=[32],
+                            client_nodes=[128, 256, 512],
+                            iterations=20,
+                            tensor_bytes=[1024, 8192, 16384, 32769, 65538,
+                                          131076, 262152, 524304, 1024000,
+                                          2048000],
+                            tensors_per_dataset=[1,2,4],
+                            client_threads=[1,2,4,8,16,32]):
+
+        """Run the data aggregation scaling tests
```
Comment (ashao) [inline driver.py:320]: I'd maybe just mention real quick that the actual tests that will be run are a permutation of: client_nodes, clients_per_node, tensor_bytes, tensors_per_dataset, client_threads.

**Review by ashao — CHANGES_REQUESTED** [review]
Just failry minor comments. The main one that we might want to sync on briefly is how many `cpus_per_task` we should be setting so that we can make sure we get accurate results on machines that have hypthreading on or off.

File: cpp-data-aggregation/aggregation_consumer.cpp (hunk 2 above)
Comment (mellis13) [reply cpp-data-aggregation/aggregation_consumer.cpp:42]: Agreed on wanting to sync for each iteration.  I think the ``MPI_Barrier`` below where we check that the list is empty will synchronize all of the ranks before any new datasets are added to the aggregation list and datasets are pulled from the aggregation list.  I think the extra ``MPI_Barrier`` might just add extra time to the loop without more sync.

Comment (ashao) [reply cpp-data-aggregation/aggregation_consumer.cpp:42]: Ah yeah ok, I see that. I agree no need for `MPI_Barrier`

**Review by ashao — APPROVED** [review]
Looks good. Thanks for getting that in!

**Pull request merged**